manta-crypto/src/eclair/cmp.rs

```diff
@@ -56,30 +60,155 @@ where
     }
 }
 
-/* TODO:
-impl<T, Rhs> PartialEq<Rhs> for T
+/// Implements [`PartialEq`] for the given `$type`.
+macro_rules! impl_partial_eq {
+    ($($type:tt),* $(,)?) => {
+        $(
+            impl<Rhs> PartialEq<Rhs> for $type
+            where
+                $type: cmp::PartialEq<Rhs>,
+            {
+                #[inline]
+                fn eq(&self, rhs: &Rhs, _: &mut ()) -> bool {
+                    cmp::PartialEq::eq(self, rhs)
+                }
+
+                #[inline]
+                fn ne(&self, rhs: &Rhs, _: &mut ()) -> bool {
+                    cmp::PartialEq::ne(self, rhs)
+                }
+            }
+        )*
+    };
+}
+
+impl_partial_eq!(bool, u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
+
+impl<T, Rhs, COM> PartialEq<Vec<Rhs>, COM> for Vec<T>
+where
+    COM: Has<bool>,
+    Bool<COM>: Constant<COM, Type = bool> + BitAnd<Bool<COM>, COM, Output = Bool<COM>>,
+    T: PartialEq<Rhs, COM>,
+{
+    #[inline]
+    fn eq(&self, rhs: &Vec<Rhs>, compiler: &mut COM) -> Bool<COM> {
+        if self.len() != rhs.len() {
+            false.as_constant(compiler)
+        } else {
+            let mut are_equal = true.as_constant::<Bool<COM>>(compiler);
+            for (lhs, rhs) in self.iter().zip(rhs) {
+                are_equal = are_equal.bitand(lhs.eq(rhs, compiler), compiler);
+            }
+            are_equal
+        }
+    }
+
+    #[inline]
+    fn assert_equal(&self, rhs: &Vec<Rhs>, compiler: &mut COM)
+    where
+        COM: Assert,
+    {
+        if self.len() != rhs.len() {
+            let not_equal = false.as_constant(compiler);
+            compiler.assert(&not_equal);
+        } else {
+            for (lhs, rhs) in self.iter().zip(rhs) {
+                compiler.assert_eq(lhs, rhs);
+            }
+        }
```

Comment:
Can we simply use `fn eq(&self, rhs: &Vec<Rhs>, compiler: &mut COM) -> Bool<COM>` and assert on its output?

Reply:
This is more efficient from the point of view of # of constraints. This is the entire purpose of the `assert_equal` function. See the docstring.